### PR TITLE
Allow multiple storages

### PR DESCRIPTION
**Pull request by andrewkav**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -27,6 +36,23 @@ func execute() error {
 
 	viberKey := os.Getenv("VIBER_KEY")
 	callback_URL := os.Getenv("CALLBACK_URL")
+
+	// setup persistence
```
Comment (gburanov) [inline main.go:40]: ufff it took me some time to understand it. As far as I understand

* You create a global pointer to function
* You init it first with `newPQUserDAO` postgres DAO
* Then if u have a redis u create `newRedisUserDAO` and override function pointer.
* And then u use this function in just one place - initStorage()

Why can't we do simplier here? For example just tune the newStorage

```
func newStorage() (*Storage, error) {
	var err error
	var dao userDAO
	if redisAddr := os.Getenv("REDIS_ADDR"); redisAddr != "" {
		dao = newRedisUserDAO(redis.NewClient(&redis.Options{Addr: redisAddr}))
	} else {
		dao, err = newPQUserDAO()
		if err != nil {
			return nil, err
		}
	}

	return &Storage{
		//users:      sync.Map{},
		userDAO: dao,
	}, nil
}
```

the pointer initialization in init and overriding later just makes code hard to follow IMHO. WDYT?

File: main.go [hunk 2]
```diff
@@ -27,6 +36,23 @@ func execute() error {
...
+	// prefer redis persistence over SQL
+	if redisAddr := os.Getenv("REDIS_ADDR"); redisAddr != "" {
+		var (
+			rudOnce sync.Once
+			rud     *redisUserDAO
+		)
+		newUserDAO = func() (userDAO, error) {
+			rudOnce.Do(func() {
```
Comment (gburanov) [inline main.go:48]: this is executed in one thread - before we spawn threads, so imho runOnce is not needed here

File: storage.go [hunk 3]
```diff
@@ -5,20 +5,31 @@ import (
 	"log"
 )
 
+var (
+	newUserDAO func() (userDAO, error)
+)
+
+type userDAO interface {
+	load(id string) (*StorageUser, error)
```
Comment (gburanov) [inline storage.go:13]: Thank you very much! I like the idea making the interface for user persistence

**Comment by gburanov** [comment]
+ there is one global point here - so now Redis is implemented just INSTEAD of postgres.

They use the same interface, and exchangeable.

This is nice, but do we want it? I mean, redis is even worse the postgres in the terms of data persistence - it is in memory storage, and data usually has TTL - and removed after that point https://redis.io/commands/ttl

So having redis instead not postgres - I am not sure =)

Maybe we better use it in addition to postgres? I see 2 options here: easy and hard.

* Easy - use redis as additional layer. So when we save - we save both to redis and to postgres
* When we load  - we load from redis, and only if not found - from postgres.
* The good part is that it is easy to do - you have interfaces already
* The bad part - if things go out of sync - it is not nice.

Hard way - use WAL to ease load on postgres. So when we write in postgres and it is overloaded - we write to redis first, and then later write to postgres

File: main.go (hunk 1 above)
Comment (andrewkav) [reply main.go:40]: I didn't want to change to much code and have IoC :) i think we can make it better and move initializtion to main function and inject the dependecies :) will preapare another PR

**Comment by andrewkav** [comment]
> This is nice, but do we want it? I mean, redis is even worse the postgres in the terms of data persistence - it is in memory storage, and data usually has TTL - and removed after that point https://redis.io/commands/ttl
TTL isn't a concern, we set TTL to never expire, persistence wise - redis has AOF (append only file) persistence no data will be lost - https://redis.io/topics/persistence
I don't see any problems using redis tbh.
> Hard way - use WAL to ease load on postgres. So when we write in postgres and it is overloaded - we write to redis first, and then later write to postgres
i don't follow

File: main.go (hunk 2 above)
Comment (andrewkav) [reply main.go:48]: will remove according to my previous comment :)
